src/compiler/parser.ts

```diff
@@ -3480,7 +3518,7 @@ module ts {
                 case SyntaxKind.VoidKeyword:
                     return parseVoidExpression();
                 case SyntaxKind.LessThanToken:
-                    return parseTypeAssertion();
+                    return (sourceFile.jsxFactory && tryParse(() => parseJSXElement(true))) || parseTypeAssertion();

```

Comment:
At first I had a preprocessing phase that was building a map of closing tags to improve performance like you described.
The problem with that approach is that it produces very unfriendly errors in case of typo in closing tag.
I so then switched to a simple array of closing tag positions, and I checked in case of non self closing element if there was at least one closing tag later in the file. 
However I thought that this optimisation was complex for very few results in this PR since I only parse JSX elements if the file contains the triple slash directive. (and that there will certainly be closing tags in this case)